File: projects/melon-runtime/MelonRuntime.Core/Entities/Event.cs
using Jint.Native;
using MelonRuntime.Abstractions.Generic;
using System.Collections.Concurrent;

namespace MelonRuntime.Core.Entities
{
    public enum EventCaller
    {
        System,
        Cronjob,
        User
    }

    public enum EventType
    {
        LocalInputOutput,
        RemoteInputOutput,
        InteropTransaction
    }

    public class Event
    {
        public IReadOnlyCollection<string> Metadata { get; set; }
        public EventCaller Caller { get; set; }
        public EventType Type { get; set; }
        public DateTimeOffset CreatedAt { get; set; }
        public DateTimeOffset UpdatedAt { get; set; }
        public bool Finished { get; set; }
        public bool Paused { get; set; }

        private readonly IMelon<JsValue> _melon;
        private readonly LinkedList<JsValue> _actions;

        public Event(
            IMelon<JsValue> melon,
            EventCaller caller,
            EventType type,
            JsValue[] actions)
        {
            _melon = melon;
            _actions = new LinkedList<JsValue>(actions);

            Metadata = new ConcurrentBag<string>();
            Caller = caller;
            Type = type;
        }

        public async Task Run()
        {
            if (Finished || _actions.Count == 0)
            {
                Finished = true;
                return;
            }

            if (!Paused)
            {
                await Task.Run(() => _melon.InteropInvoke(_actions.First()));
                _actions.RemoveFirst();
            }

            await Run();
        }
    }
}

File: projects/melon-core/src/partials/modules/runtime/Event.ts
import { _crypto } from "../../statics/_crypto";
import { _log } from "../console/_log";

enum EventCaller {
    System,
    Cronjob,
    User
}

enum EventType {
    LocalInputOutput,
    RemoteInputOutput,
    InteropTransaction
}

class EventChain {
    private static events: Map<string, Event>;

    static startNew(caller: EventCaller, type: EventType, actions: Function[]) {
        const name = _crypto.randomUUID();
        const event = new Event(name, caller, type, actions);
        this.events.set(name, event);
        this.events.get(name).run();
    }

    static getEvents(condition: Function = null) {
        if (!condition) {
            return this.events;
        }
        
        const resultMap = new Map<string, Event>();

        for (const event of this.events) {
            if (condition(event[1])) {
                resultMap.set(event[0], event[1]);
            }
        }

        return resultMap;
    }
}

class Event {
    private name: string;
    private actions: Function[];
    private type: EventType;
    private caller: EventCaller;
    private finished: boolean;

    get isFinished() {
        return this.finished;
    }

    get eventName() {
        return this.name;
    }

    get eventType() {
        return this.type;
    }

    get eventCaller() {
        return this.caller;
    }

    constructor(name: string, caller: EventCaller, type: EventType, actions: Function[]) {
        this.name = name;
        this.actions = actions;
        this.type = type;
        this.caller = caller;
    }

    async run(): Promise<void> {
        if (!this.actions.length || this.finished) {
            return;
        }

        const nextAction = this.actions.shift();
        const promise = new Promise(resolve => resolve(nextAction()));
        await promise;
        await this.run();
    }

    finish() {
        this.finished = true;
    } 
}

export { Event, EventCaller, EventType, EventChain }